### PR TITLE
add ability to generate chart schema.json

### PR DESCRIPTION
**Pull request by albertlockett**
Adds the ability to generate the json schema for a chart. Users will do this by adding the `JsonSchema` field to the chart definition as shown here:

https://github.com/solo-io/skv2/blob/33b1815a064e71bd033f148d63c323d5e60d95e4/codegen/cmd_test.go#L309

BOT NOTES: 
resolves https://github.com/solo-io/gloo-mesh-enterprise/issues/7186

### CODE REVIEW

File: codegen/render/render.go [hunk 1]
```diff
@@ -12,8 +12,13 @@ type Resource = model.Resource
...
+}
+
 type OutFile struct {
-	Path       string
-	Permission os.FileMode
-	Content    string // set by Renderer
+	Path           string
+	Permission     os.FileMode
+	Content        string         // set by Renderer
+	HeaderOverride HeaderOverride // overrides the defauld the default header of the outfile writer
```
Comment (josh-pritchard) [inline codegen/render/render.go:23]: typo

File: codegen/render/funcs.go [hunk 2]
```diff
@@ -562,6 +571,341 @@ func fromJSON(str string) map[string]interface{} {
 	return m
 }
 
+// toJSONSchema generates the json schema for the values of the helm chart
+func toJSONSchema(values values.UserHelmValues) string {
+	reflector := createJsonSchemaReflector(values)
+	schema := new(jsonschema.Schema)
+	schema.Version = "https://json-schema.org/draft/2020-12/schema"
```
Comment (josh-pritchard) [inline codegen/render/funcs.go:578]: Where did this come from? Maybe a comment or link?

File: codegen/render/funcs.go [hunk 3]
```diff
@@ -562,6 +571,341 @@ func fromJSON(str string) map[string]interface{} {
...
+		}
+
+		return nil
+	}
+}
+
+// initialize a new instance of the jsonschema Reflector with configured to
+// create json schema for chart.
+func createJsonSchemaReflector(values values.UserHelmValues) *jsonschema.Reflector {
+	// we might want to pass configuration ot this from values.UserHelmValues
```
Comment (josh-pritchard) [inline codegen/render/funcs.go:700]:
```suggestion
	// we might want to pass configuration of this from values.UserHelmValues
```

File: codegen/render/funcs.go [hunk 4]
```diff
@@ -562,6 +571,341 @@ func fromJSON(str string) map[string]interface{} {
...
+	var buf bytes.Buffer
+	err := json.Indent(&buf, []byte(src), "", "  ")
+	if err != nil {
+		panic(err)
+	}
+	return buf.String()
+}
+
+// reverse mutates a slice of strings to reverse the order.
+func reverse(s []string) {
```
Comment (josh-pritchard) [inline codegen/render/funcs.go:903]: one would think golang standard lib would include this....not that I could find it.

File: codegen/writer/writer.go [hunk 5]
```diff
@@ -49,7 +49,9 @@ func (w *DefaultFileWriter) WriteFiles(files []render.OutFile) error {
 			commentPrefix = "#"
 		}
 
-		if w.Header != "" {
+		if file.HeaderOverride != nil {
```
Comment (josh-pritchard) [inline codegen/writer/writer.go:52]: nit: couldn't we disable the header based on file type? I am fine with it as is too.

**Review by josh-pritchard — COMMENTED** [review]
only a few nits: This looks fantastic and being an opt-in feature lets get this in so we can try it out

File: codegen/render/funcs.go (hunk 4 above)
Comment (albertlockett) [reply codegen/render/funcs.go:903]: yeah I googled for quite a while and I couldn't find something that does it :(

File: codegen/writer/writer.go (hunk 5 above)
Comment (albertlockett) [reply codegen/writer/writer.go:52]: yeah for sure good call, I agree w/ that in hindsight

At the time I thought this HeaderOverride solution might be more flexible but given that this one json file is the only use case and the added complexity I'm inclined to just disable header for json as well. it aligns better w/ the current pattern of choosing header by file type.

Comment (albertlockett) [reply codegen/writer/writer.go]: fixed, I removed this

File: codegen/render/funcs.go (hunk 2 above)
Comment (albertlockett) [reply codegen/render/funcs.go]: fixed, I got this from the jsonschema package.

File: codegen/render/render.go (hunk 1 above)
Comment (albertlockett) [reply codegen/render/render.go]: fixed I removed this.

File: codegen/render/funcs.go (hunk 4 above)
Comment (fernandosanchezs) [reply codegen/render/funcs.go:903]: `Implementation is left to the reader.`

Comment (puertomontt) [reply codegen/render/funcs.go:903]: wow

**Review by josh-pritchard — DISMISSED** [review]
LGTM! Great work!

**Review by puertomontt — APPROVED** [review]

**Review by josh-pritchard — APPROVED** [review]

**Pull request merged**